yktracec/src/promote.rs

```diff
@@ -0,0 +1,75 @@
+//! Software constant recorder.
+//!
+//! This is used for promoting variables to constants when compiling a trace.
+
+use std::{cell::RefCell, collections::VecDeque};
+
+// FIXME: Try to move this into ykrt in the existing thread local.
+thread_local! {
+    /// The thread's value recorder.
+    ///
+    /// This has to be specific to a thread to avoid cross-talk from other threads which may be
+    /// executing functions with promoted arguments at the same time.
+    pub static VAL_REC: RefCell<ValueRecorder> = RefCell::new(ValueRecorder::default());
+}
+
+/// A value recorder observes and records constant values during tracing. The trace compiler
+/// queries this when building a trace to replace promoted variables with the observed constants.
+#[derive(Debug, Default)]
+pub struct ValueRecorder {
+    // `true` when recording promotions.
+    record_enable: bool,
+    // A vector of observed values of to-be-promoted variables, in the order their `yk_promote()`
```

Comment:
"A vector of promoted values in the order their `yk_promote()` calls were encountered in the trace"?